### PR TITLE
feat(resharding): flat storage resharding deltas

### PR DESCRIPTION
**Pull request by Trisfald**
Implementation of flat storage deltas split for flat storage resharding.

Before this change, only flat storage key pairs at flat head height were split during resharding. Now flat storage deltas are part of the splitting process as well. 
On a high level, it works by concatenating the iterator over flat storage and the iterator over deltas for all block heights between chain head an flat storage head.

Part of #12174

### CODE REVIEW

File: chain/chain/src/flat_storage_resharder.rs [hunk 1]
```diff
@@ -396,8 +405,74 @@ impl FlatStorageResharder {
...
+                .iter(*shard_uid)
+                // Get the flat storage iter and wrap the value in Optional::Some to
+                // match the delta iterator so that they can be chained.
+                .map_ok(|(key, value)| (key, Some(value)))
+                // Wrap the iterator's item into an Entry.
+                .map(|entry| FlatStorageAndDeltaIterItem::Entry(entry)),
+        );
+
+        // Get all the blocks from flat head to the wanted block hash.
+        let flat_storage = self
```
Comment (wacban) [inline chain/chain/src/flat_storage_resharder.rs:428]: I'm a bit confused about having flat_store as an argument and getting flat_storage from runtime here. Why do you need both and how do they differ? Is there any way to make it simpler? Ideally there would only be one flat `thing` or they would be clearly different and both would be obtained same way.

File: chain/chain/src/flat_storage_resharder.rs [hunk 2]
```diff
@@ -396,8 +405,74 @@ impl FlatStorageResharder {
...
+            let deltas = flat_store.get_delta(*shard_uid, block).map_err(|err| {
+                StorageError::StorageInconsistentState(format!(
+                    "can't retrieve deltas for flat storage at {block}/{shard_uid:?}({err})"
+                ))
+            })?;
+            let Some(deltas) = deltas else {
+                continue;
+            };
+            // Chain the iterators effectively adding a block worth of deltas.
+            // Before doing so insert a commit point to separate changes to the same key in different transactions.
```
Comment (wacban) [inline chain/chain/src/flat_storage_resharder.rs:454]: Nice, I wouldn't have thought of that. How did you find that it's needed?

File: chain/chain/src/flat_storage_resharder.rs [hunk 3]
```diff
@@ -507,29 +582,29 @@ fn copy_kv_to_child(
...
     status: &SplittingParentStatus,
     key: Vec<u8>,
-    value: FlatStateValue,
+    value: Option<FlatStateValue>,
     store_update: &mut FlatStoreUpdateAdapter,
 ) {
-    store_update.set(status.left_child_shard, key.clone(), Some(value.clone()));
-    store_update.set(status.right_child_shard, key, Some(value));
+    store_update.set(status.left_child_shard, key.clone(), value.clone());
+    store_update.set(status.right_child_shard, key, value);
```
Comment (wacban) [inline chain/chain/src/flat_storage_resharder.rs:597]: not related to this PR
It feels like the set method should only need a reference to the key and value. In theory it should serialize those and for that ref should be sufficient. Then the clone would not be needed.

File: chain/chain/src/flat_storage_resharder.rs [hunk 4]
```diff
@@ -980,11 +1066,193 @@ mod tests {
...
+        // Buffered receipt '0' should exist with updated value only in the left child.
+        assert_eq!(
+            flat_store.get(left_child_shard, &buffered_receipt_0_key.to_vec()),
+            Ok(buffered_receipt_0_value_1.map(|val| FlatStateValue::inlined(&val)))
+        );
+        assert_eq!(flat_store.get(right_child_shard, &buffered_receipt_0_key.to_vec()), Ok(None));
+        // Buffered receipt '1' shouldn't exist.
+        for child in [left_child_shard, right_child_shard] {
+            assert_eq!(flat_store.get(child, &buffered_receipt_1_key.to_vec()), Ok(None));
+        }
```
Comment (wacban) [inline chain/chain/src/flat_storage_resharder.rs:1255]: Ultimately we should check that the state root matches that on chain (which was calculated from memtrie resharding). I'm not sure how easy would it be to do in test though. We should definitely do it when replacing the temporary memtrie with new one though.

File: chain/chain/src/flat_storage_resharder.rs [hunk 5]
```diff
@@ -396,8 +405,74 @@ impl FlatStorageResharder {
...
+        Ok(iter)
+    }
+}
+
+/// Enum used to wrap the `Item` of iterators over flat storage contents or flat storage deltas. Its
+/// purpose is to insert a marker to force store commits during iteration over all entries. This is
+/// necessary because otherwise deltas might set again the value of a flat storage entry inside the
+/// same transaction.
+enum FlatStorageAndDeltaIterItem {
+    Entry(Result<(Vec<u8>, Option<FlatStateValue>), FlatStorageError>),
```
Comment (wacban) [inline chain/chain/src/flat_storage_resharder.rs:470]: nit: Is there any chance to remove the error from here? Would it be nicer?

**Review by wacban — APPROVED** [review]
LGTM

I really like how few changes you needed to the original code, the delta handling fits like a glove ;)

**Comment by wacban** [comment]
Just out of curiosity how much of it is copy pasted from resharding V2 (RIP)?

File: chain/chain/src/flat_storage_resharder.rs (hunk 2 above)
Comment (Trisfald) [reply chain/chain/src/flat_storage_resharder.rs:454]: Tests failed with transaction overwrite error 😉

**Comment by Trisfald** [comment]
> Just out of curiosity how much of it is copy pasted from resharding V2 (RIP)?

The idea is the same (chaining iterators). In practice though I had to code the solution differently, because resharding-v2 works on trie changes and I'm working directly on flat storage key value pairs.

File: chain/chain/src/flat_storage_resharder.rs (hunk 1 above)
Comment (Trisfald) [reply chain/chain/src/flat_storage_resharder.rs:428]: Yes, it's a bit weird. I need `flat_store` passed to the function due to lifetimes on the iterator - this prevents me from instantiating a `flat_store` inside the `flat_storage_iterator`.

And then `flat_storage` and `flat_store`, although they have similar name, are totally independent. I couldn't get one from the other

File: chain/chain/src/flat_storage_resharder.rs (hunk 5 above)
Comment (Trisfald) [reply chain/chain/src/flat_storage_resharder.rs:470]: The only way I think would be to `map` the error into a `panic`, but that's.. brutal

**Pull request merged**